### DISCUSSION TITLE
Support for disabling any kind of entity mapping and caching

### DISCUSSION THREAD

**Discussion by francescortiz** [mikro-orm/mikro-orm#4653] [Ideas]
**Is your feature request related to a problem? Please describe.**
We use `em.fork()` compulsively for the following reasons:

1. We work with large ammounts of data.
2. In consequence, we carefully craft our queries, and we never query duplicate data in our processes.
3. The entity mapper is highly innefficent with large datasets. When we discovered this, I reduced a process from more than one day  (i didn't let it finish) to just 15 minutes just by doing `em.fork()` on every loop iteration.
4. Also, according to the docs, `em.upsert` has different behavior depending on wether the entity mapper has seen the entity or not, which makes it hard to predict how it willl behave. Doc quoted below:

> https://mikro-orm.io/docs/entity-manager
> If the entity is already present in current context, there won't be any queries - instead, the entity data will be assigned and an explicit flush will be required for those changes to be persisted.

**Describe the solution you'd like**
It would be great if we could totally disable the entitiy mapper and/or any other kind of peristence managed by mikro-orm beyond the operation we are executing.

**Describe alternatives you've considered**
For now we have decided to use query builders as much as possible because they work in a more comprehensive way, while still giving use the power of relational mapping of the database (code is easier and safer to write). 

**Additional context**
- We use transactions when we need atomicity.
- We love the relational mapping of queries and results that `mikro-orm` gives us, so we don't consider using alternatives.

**Comment by B4nan** [comment]
This is very abstract, you will need to be much more concrete if we should take this forward. Let's start with what you mean by "entity mapper", you mean the hydration layer?

> The entity mapper is highly innefficent with large datasets. When we discovered this, I reduced a process from more than one day (i didn't let it finish) to just 15 minutes just by doing em.fork() on every loop iteration.

If this is a performance problem, you should start with a complete reproduction for that, so I can understand what problem are you dealing with.

edit: moving to discussions before we settle on what should be done

**Comment by B4nan** [comment]
Btw I guess you know about the `disableIdentityMap` option?

https://mikro-orm.io/docs/entity-manager#disabling-identity-map-and-change-set-tracking

**Reply by francescortiz** [reply]
Thanks for the guidance. I didn't know exactly where to point to int his topic. I didn't know about the `disableIdentityMap` and it looks exactly as the behavior that I look for. Is there a way to make it the default? Thank you

**Reply by B4nan** [reply]
Now you can: https://github.com/mikro-orm/mikro-orm/commit/ef32b14111245caba8cc9f7298eb8f81a88488a6

**Reply by francescortiz** [reply]
Wow, this is great. Thank you very much!

**Comment by francescortiz** [comment]
After revieweing the source code a bit, I am wondering that what we need might be `em.clear();` because I don't know how `disableIdentityMap: true` affects transactional entity managers. Is `disableIdentityMap: true` making queries inside the transactional `em`'s transaction or is it making the queries outside the transaction? Thank you very much

**Reply by B4nan** [reply]
The context is cleared automatically when you use this flag:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/EntityManager.ts#L152-L159

This option is about reading, so `em.find` and similar, and that itself is not automatically wrapped in any transaction.

**Comment by yackinn** [comment]
Is there any option to disable identity map in query builder as well?

**Reply by B4nan** [reply]
Just fork first before you create it (or use `qb.execute`), the disable identity map option is doing exactly that behind the scenes.

**Reply by yackinn** [reply]
Is it possible to query data and get entity instances in return and disabling the identity map?

```ts
em.qb().select().getResult() // instances
em.qb().select().execute() // objects

// we need this with entity instances
em.qb().select().execute()
```

**Reply by yackinn** [reply]
> Just fork first before you create it (or use `qb.execute`), the disable identity map option is doing exactly that behind the scenes.

Also another question about the identity map when inserting data.
```ts
await em
      .createQueryBuilder(Entity)
      .insert(data)
      .execute();
em.clear(); // is this unnecessary then as the identity map is disabled when using execute anyway?
```

**Reply by B4nan** [reply]
As I said, just fork first.

```ts
em.fork().qb().select().getResult() // instances
```

> em.clear(); // is this unnecessary then as the identity map is disabled when using execute anyway?

`qb.execute` doesnt touch identity map anyhow, so yes, this call is not necessary, and can be actually harmful if there was something else in the context. you should almost never use `em.clear` outside of tests.

**Reply by yackinn** [reply]
is there any way to get back detached entity instances when using `qb().select("*").execute()`? If I understand correctly `em.map(...)` will add the created instances to the identity map.

**Reply by B4nan** [reply]
there are no entity instances with `qb.execute`, so you cant "get anything back". yes, you can use `em.map` to create entities from raw data.

**Reply by B4nan** [reply]
oh now i get it. again, same principle, fork first.

**Reply by yackinn** [reply]
thank you

**Reply by yackinn** [reply]
Just to make sure. Running
```ts
const insertedData = await qb
      .insert(update)
      .returning('id')
      .execute();
```
won't add to the `identityMap` even without `.clone()` right?

**Reply by B4nan** [reply]
Correct, or if you want to map to entities, just do `em.fork().qb(...)`.

**Reply by yackinn** [reply]
Every `.fork()` creates another db connection right? If we do that too many times, we too many db connections, so we'd rather not to. Mapping to entities without adding it to the `identityMap` is not possible right now, correct?

**Reply by B4nan** [reply]
Nope, forks have nothing to do with connections directly, transaction context is what holds a connection, not a fork on its own.

**Reply by yackinn** [reply]
ok, thank you
